lib/deimos/utils/schema_class.rb

```diff
@@ -10,10 +10,23 @@ class << self
         # @return [Array<String>]
         def modules_for(namespace)
           modules = ['Schemas']
-          namespace_folder = namespace.split('.').last
-          if Deimos.config.schema.generate_namespace_folders && namespace_folder
-            modules.push(namespace_folder.underscore.classify)
+
+          if Deimos.config.schema.generate_namespace_folders
+            namespace_map = Deimos.config.schema.schema_namespace_map.sort_by{|k,v| -k.length}.to_h
+            namespace_override = namespace_map.keys.detect{ |k| namespace.include?(k) }
+
+            module_namespace = if namespace_override.present?
+              # don't use Schemas default module
+              modules = Array(namespace_map[namespace_override].dup)
+              namespace.gsub(/#{namespace_override}.?/, '')
+            else
+              namespace
+            end
+
+            namespace_folders = module_namespace.split('.').map{ |f| f.underscore.camelize }
```

Comment:
Why the move from `classify` to `camelize`? This is a breaking change.

Reply:
This is because module names can be plural. If you have a bunch of log related schemas under the namespace `com.my_org.logs` you would want them in the Logs module rather than Log.

I see where you're coming from with this being a breaking change, however simply updating the gem will not break anything.

When you regenerate the classes, if you previously used the `generate_namespace_folders` option you are going to have to modify references anyway as this PR changes that to use the entire namespace.

The modules would change from the current `Schemas => Log` to `Schemas => Com => MyOrg => Logs`. You could alternatively use the new config option to keep the previous behaviour like this:

```
schema_namespace_map({
        'com.my_org.logs' => ['Schemas', 'Log']
      })
```

Reply:
Not sure I understand why the default behavior changed? (I must have missed that in the original PR.) If the point of this PR is to allow overriding the current behavior, why would we _also_ change the default?

Reply:
well currently it only uses the very last part of the namespace for modularization. For this configuration to be useful we need to use the entire namespace as one of the issues is having similarly named schemas conflicting. To give a made up example:
```
com.my-org.client.logs
com.my-org.customer.logs
com.my-other-org.customer.logs
```

All of these would currently be under Schemas => Log and would conflict. If we don't take into account the entire namespace there is no way to differentiate them or provide ways to override them.

One of the things I considered when deciding where we want our schema classes to be generated is to put them alongside the avro definitions, which must be fully namespaced. This change also would enable users of the gem to do that should they wish.

IMO this is the most flexible approach that allows users to have their generated files modularized in the way that suits their codebase

Reply:
I thought about this some more and I think we would be able to keep the current behaviour and it probably makes sense to do so. I'll see if I can get a commit together and we can see how that looks

Reply:
ok, so I have some code that will accomplish this. The caveat here is that the namespace map config will need to be updated more frequently as a result. I'll try to explain what I mean with an example:

Let's say we have the following map:
```
schema_namespace_map({
    'com.company.sub_org' => 'SchemaClasses',
    'com.company.sub_org.timekeeping' => ['SchemaClasses', 'Timekeeping']
    'com.company.sub_org.hiring' => ['SchemaClasses', 'Hiring']
  })
```

And we have schemas `com.company.sub_org.timekeeping.some-schema` and `com.company.sub_org.hiring.some-other-schema`

But now we want to add a schema for the payroll domain `com.company.sub_org.payroll.some-schema`. We're going to need to add this to the config otherwise it will be generated in the SchemaClasses directory which is not ideal.

In fact every time we add a new schema with a different sub namespace we will need to also change the config. This was what I was trying to avoid by changing the existing behaviour.

My personal preference would be to leave this as it is by changing the default to use the full namespace as it seems more of a convention over configuration approach. That said, I'm happy to change it if necessary and will wait to hear what you think

Reply:
Sorry for the delay. 

I understand the desire here. I'm wondering if we can do something like... if `schema_namespace_map` is used at all, then change the default to what you're describing. But that seems like it's too much magic. 

Could we add another config for this? I know this ends up having lots of different options around how/where schemas are generated.

Actually... could we repurpose `generate_namespace_folders` to control this instead? From what I understand you're the only folks using this option :grin: 

I think we have an option when we do a version bump to change the default behavior and possibly deprecate/remove the config. I'm thinking about doing this when we (ultimately) move from ruby-kafka/phobos to ruby-rdkafka as ruby-kafka is deprecated and unmaintained. Maybe you can add a comment to that effect.

Reply:
yes I think renaming generate_namespace_folders to use_full_namespace is the best option here - I'll add a commit to do that